// dm-util/src/index.js
// Generated by CoffeeScript 1.10.0
(function() {
  var EventEmitter, getScript, getServiceEE;

  EventEmitter = require('events').EventEmitter;

  getScript = require('./getScript');

  getServiceEE = function() {
    var base;
    if (typeof window === "undefined" || window === null) {
      return;
    }
    if (global._service == null) {
      global._service = {};
    }
    if ((base = global._service).ee == null) {
      base.ee = new EventEmitter();
    }
    return global._service.ee;
  };

  exports.getScript = function(source, cb) {
    var _listener, base, ee, scripts;
    if (!(ee = getServiceEE())) {
      return typeof cb === "function" ? cb() : void 0;
    }
    scripts = (base = global._service).scripts != null ? base.scripts : base.scripts = {
      loaded: {},
      loading: {}
    };
    if (scripts.loaded[source]) {
      return typeof cb === "function" ? cb() : void 0;
    }
    if (scripts.loading[source]) {
      if (cb == null) {
        return;
      }
      return ee.on('loadScript', _listener = function(loadedSource) {
        if (source === loadedSource) {
          ee.removeListener('loadScript', _listener);
          return cb();
        }
      });
    } else {
      scripts.loading[source] = true;
      return getScript(source, function() {
        delete scripts.loading[source];
        scripts.loaded[source] = true;
        ee.emit('loadScript', source);
        return typeof cb === "function" ? cb() : void 0;
      });
    }
  };

  exports.ajax = require('./ajax');

  exports.file = require('./file');

  exports.numbers = require('./numbers');

}).call(this);
